### ISSUE TITLE
UnsupportedOperationException: I kinna do it, cap'n!

### ISSUE THREAD

**Issue by svrlopatrik** [commonsguy/cwac-saferoom#17]
Hi,
can you help me with my problem.

I'm trying to encrypt Room with your library but getting: `java.lang.RuntimeException: java.lang.UnsupportedOperationException: I kinna do it, cap'n!` when initializing database.

This is initialization part:
```
@Provides
@ApplicationScope
    public fun providesDatabase(context: Context): Database {
        val safeHelperFactory = SafeHelperFactory.fromUser(Editable.Factory.getInstance().newEditable("secret_password"))
        return Room.databaseBuilder(context, Database::class.java, Constants.DATABASE_NAME)
                .openHelperFactory(safeHelperFactory)
                .build()
    }
```
I'm using these versions:

```
room                    = '1.1.0-alpha2'
safeRoom                = '0.3.2'

implementation "android.arch.persistence.room:runtime:$room"
annotationProcessor "android.arch.persistence.room:compiler:$room"
kapt "android.arch.persistence.room:compiler:$room"
implementation "com.commonsware.cwac:saferoom:$safeRoom"
```
Application runs on AVD with version 8.1.0.

Thank you

**Comment by commonsguy** [comment]
I need the complete Java stack trace.

**Comment by svrlopatrik** [comment]
Here it is:
```
02-24 17:23:32.057 7911-7911/sk.kotlin.skeleton E/AndroidRuntime: FATAL EXCEPTION: main                                                                  Process: sk.kotlin.skeleton, PID: 7911                                                                  java.lang.RuntimeException: Unable to start activity ComponentInfo{sk.kotlin.skeleton/sk.kotlin.skeleton.ui.activity.detail.DetailActivity}: java.lang.RuntimeException: java.lang.UnsupportedOperationException: I kinna do it, cap'n!
                                                                      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
                                                                      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
                                                                      at android.app.ActivityThread.-wrap11(Unknown Source:0)
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
                                                                      at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                      at android.os.Looper.loop(Looper.java:164)
                                                                      at android.app.ActivityThread.main(ActivityThread.java:6494)
                                                                      at java.lang.reflect.Method.invoke(Native Method)
                                                                      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
                                                                   Caused by: java.lang.RuntimeException: java.lang.UnsupportedOperationException: I kinna do it, cap'n!
                                                                      at sk.kotlin.skeleton.viewmodel.ViewModelFactory.create(ViewModelFactory.kt:29)
                                                                      at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:133)
                                                                      at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:101)
                                                                      at sk.kotlin.skeleton.ui.activity.BaseActivity.onCreate(BaseActivity.kt:35)
                                                                      at android.app.Activity.performCreate(Activity.java:6999)
                                                                      at android.app.Activity.performCreate(Activity.java:6990)
                                                                      at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
                                                                      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
                                                                      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
                                                                      at android.app.ActivityThread.-wrap11(Unknown Source:0) 
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
                                                                      at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                      at android.os.Looper.loop(Looper.java:164) 
                                                                      at android.app.ActivityThread.main(ActivityThread.java:6494) 
                                                                      at java.lang.reflect.Method.invoke(Native Method) 
                                                                      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
                                                                   Caused by: java.lang.UnsupportedOperationException: I kinna do it, cap'n!
                                                                      at com.commonsware.cwac.saferoom.Helper.setWriteAheadLoggingEnabled(Helper.java:108)
                                                                      at android.arch.persistence.room.RoomDatabase.init(RoomDatabase.java:113)
                                                                      at android.arch.persistence.room.RoomDatabase$Builder.build(RoomDatabase.java:612)
                                                                      at sk.kotlin.skeleton.di.modules.DatabaseModule.providesDatabase(DatabaseModule.kt:24)
                                                                      at sk.kotlin.skeleton.di.modules.DatabaseModule_ProvidesDatabaseFactory.get(DatabaseModule_ProvidesDatabaseFactory.java:24)
                                                                      at sk.kotlin.skeleton.di.modules.DatabaseModule_ProvidesDatabaseFactory.get(DatabaseModule_ProvidesDatabaseFactory.java:10)
                                                                      at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
                                                                      at sk.kotlin.skeleton.di.modules.DatabaseModule_ProvidesUserDaoFactory.get(DatabaseModule_ProvidesUserDaoFactory.java:24)
                                                                      at sk.kotlin.skeleton.di.modules.DatabaseModule_ProvidesUserDaoFactory.get(DatabaseModule_ProvidesUserDaoFactory.java:10)
                                                                      at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
                                                                      at sk.kotlin.skeleton.viewmodel.DetailViewModel_Factory.get(DetailViewModel_Factory.java:23)
                                                                      at sk.kotlin.skeleton.viewmodel.DetailViewModel_Factory.get(DetailViewModel_Factory.java:9)
                                                                      at sk.kotlin.skeleton.viewmodel.ViewModelFactory.create(ViewModelFactory.kt:27)
                                                                      at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:133) 
                                                                      at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:101) 
                                                                      at sk.kotlin.skeleton.ui.activity.BaseActivity.onCreate(BaseActivity.kt:35) 
                                                                      at android.app.Activity.performCreate(Activity.java:6999) 
                                                                      at android.app.Activity.performCreate(Activity.java:6990) 
                                                                      at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214) 
                                                                      at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731) 
                                                                      at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
                                                                      at android.app.ActivityThread.-wrap11(Unknown Source:0) 
                                                                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
                                                                      at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                      at android.os.Looper.loop(Looper.java:164) 
                                                                      at android.app.ActivityThread.main(ActivityThread.java:6494) 
                                                                      at java.lang.reflect.Method.invoke(Native Method) 
                                                                      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

**Comment by commonsguy** [comment]
You are using `1.1.0-alpha2`. I am not supporting that yet.

You can try using `setJournalMode(JournalMode.TRUNCATE)` on your `RoomDatabase.Builder`. `1.1.0-alpha2` seems to default to write-ahead logging (WAL). I cannot support WAL without changes to SQLCipher itself, and I have no timetable on when that will occur.

Or, you can drop back to `1.0.0` of Room.

**Comment by devd360** [comment]
How can i work with room 1.1.0.

**Comment by nachtien** [comment]
Hi @commonsguy, I'm running into this issue also, but with Room 1.1.0.  You say that it should be working on 1.1.0, but it's not for me.  Am I doing something wrong?

Build.grade:
```
    implementation "android.arch.persistence.room:runtime:1.1.0"
    kapt "android.arch.persistence.room:compiler:1.1.0"
    implementation "com.commonsware.cwac:saferoom:0.4.0"
    implementation "net.zetetic:android-database-sqlcipher:3.5.9@aar"
```

Database.kt
```
private val pooFart = "12345".toCharArray()

@Database(
    entities = [SensorTagEntity::class, NordicEntity::class, MangoEntity::class],
    version = 1,
    exportSchema = false
)
abstract class PeripheralDatabase : RoomDatabase() {

    abstract fun sensorTagDao(): SensorTagDao
    abstract fun nordicDao(): NordicDao
    abstract fun mangoDao(): MangoDao

    companion object {
        private const val NAME = "peripherals.db"
        var factory = SafeHelperFactory(pooFart)

        fun createInMemoryDatabase(context: Context): PeripheralDatabase =
            Room.inMemoryDatabaseBuilder(
                context.applicationContext,
                PeripheralDatabase::class.java
            ).openHelperFactory(factory).setJournalMode(TRUNCATE)
                .build()

        fun createPersistentDatabase(context: Context): PeripheralDatabase =
            Room.databaseBuilder(
                context.applicationContext,
                PeripheralDatabase::class.java,
                NAME
            ).openHelperFactory(factory).setJournalMode(TRUNCATE)
                .build()
    }

    override fun clearAllTables() {
        // not implemented
    }
}
```

Exception
```
05-16 13:31:27.712 26488-26524/com.mangolabs.app E/AndroidRuntime: FATAL EXCEPTION: ForkJoinPool.commonPool-worker-1
Process: com.mangolabs.app, PID: 26488
java.lang.UnsupportedOperationException: I kinna do it, cap'n!
at com.commonsware.cwac.saferoom.Helper.setWriteAheadLoggingEnabled(Helper.java:108)
at android.arch.persistence.room.RoomDatabase.init(RoomDatabase.java:114)
at android.arch.persistence.room.RoomDatabase$Builder.build(RoomDatabase.java:631)
at com.mango.app.repository.database.PeripheralDatabase$Companion.createPersistentDatabase(PeripheralDatabase.kt:50)
at com.mango.app.repository.repository.RepositoryFactory$databasePersistent$2.invoke(RepositoryFactory.kt:17)
at com.mango.app.repository.repository.RepositoryFactory$databasePersistent$2.invoke(RepositoryFactory.kt:12)
at kotlin.SynchronizedLazyImpl.getValue(Lazy.kt:131)
at com.mango.app.repository.repository.RepositoryFactory.getDatabasePersistent(Unknown Source:25)
at com.mango.app.repository.repository.RepositoryFactory.create(RepositoryFactory.kt:31)
at com.mango.app.peripheralmanager.PeripheralManager.createRepository(PeripheralManager.kt:21)
at com.mango.app.features.scan.ScanActivity$onConnectClicked$1.doResume(ScanActivity.kt:241)
at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:54)
at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run(Dispatched.kt:161)
at kotlinx.coroutines.experimental.DispatchedContinuation.run(Dispatched.kt:25)
at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1412)
at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:285)
at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1152)
at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1990)
at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1938)
at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by commonsguy** [comment]
No, I messed something up somewhere. Working on it now. Sorry!

**Comment by commonsguy** [comment]
I was focused on the `SupportSQLiteDatabase` API. I forgot that `SupportSQLiteOpenHelper` has its *own* WAL stuff, and that I wasn't calling through to the `SupportSQLiteDatabase` implementation. Room uses the `SupportSQLiteOpenHelper` API. Working on a fix.

**Comment by commonsguy** [comment]
I integrated the solution outlined by [plackemacher in a PR](https://github.com/commonsguy/cwac-saferoom/pull/21), plus added some Room-specific tests to try to avoid this mistake in the future. This is released as v0.4.1.

**Comment by nachtien** [comment]
thanks for the quick work!
